### PR TITLE
switch causal mask for is_causal flag

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary: switch causal mask for is_causal flag

Test Plan: sandcastle & github

Differential Revision: D42089340

### CODE REVIEW

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -2305,10 +2305,14 @@ def test_aot_autograd_symbolic_exhaustive(self, device, dtype, op):
                                 # of a tracing tensor with aten._local_scalar_dense.default -
                                 # erroring out! It's likely that this is caused by data-dependent
                                 # control flow or similar.
+    torch.nn.TransformerEncoder,  # DataDependentOutputException: aten.equal compares a mask input
```
Comment (bdhirsh) [inline test/functorch/test_aotdispatch.py:2308]: I saw on the original stacktrace that the failing test was named `test_aot_autograd_module_exhaustive_nn_Transformer_cpu_float32` (https://github.com/pytorch/pytorch/actions/runs/3773547267/jobs/6415150476)

Maybe try putting `torch.nn.Transformer` instead?

Comment (drisspg) [reply test/functorch/test_aotdispatch.py:2304]: Do we know what the perf implications are for doing this in the non-causal path. I.e Does adding the torch.equal hurt the performance of a TransformerEncoder model that is currently torch.compilable?

Comment (mikekgfb) [reply test/functorch/test_aotdispatch.py:2304]: Would it not remain compilable, but with a break.  So, I'll assert that we have to choose sides here, and causal is a massive preponderance of use cases?  We might provide a way to suppress this (much like offer users ways to suppress using nested tensors, or the alignment mask correctness test during NT creation)?

File: torch/testing/_internal/common_modules.py [hunk 2]
```diff
@@ -1399,6 +1399,15 @@ def module_inputs_torch_nn_LSTM(module_info, device, dtype, requires_grad, train
                skips=(
                    DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),),
                supports_gradgrad=False),
+    # TransformerEncoder takes the same inputs as TransformerEncoderLayer
+    ModuleInfo(torch.nn.TransformerEncoder,
+               train_and_eval_differ=True,
```
Comment (albanD) [inline torch/testing/_internal/common_modules.py:1404]: In which case does training and evaluation differs?

File: torch/testing/_internal/common_modules.py [hunk 3]
```diff
@@ -1399,6 +1399,15 @@ def module_inputs_torch_nn_LSTM(module_info, device, dtype, requires_grad, train
...
+               module_inputs_func=module_inputs_torch_nn_TransformerEncoderLayer,
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:1405]: Note that using the same `module_inputs_func` for `TransformerEncoder` won't work since `TransformerEncoderLayer` and `TransformerEncoder` are constructed differently. The `module_inputs_func` specifies both constructor args and forward pass args.

File: torch/testing/_internal/common_modules.py [hunk 4]
```diff
@@ -1399,6 +1399,15 @@ def module_inputs_torch_nn_LSTM(module_info, device, dtype, requires_grad, train
                skips=(
                    DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),),
                supports_gradgrad=False),
+    # TransformerEncoder takes the same inputs as TransformerEncoderLayer
+    ModuleInfo(torch.nn.TransformerEncoder,
```
Comment (jbschlosser) [inline torch/testing/_internal/common_modules.py:1403]: Here, this should work for you:

```python
def module_inputs_torch_nn_TransformerEncoder(module_info, device, dtype, requires_grad, training, **kwargs):
    # Reuse the TransformerEncoderLayer samples since the forward args are nearly the same.
    for layer_module_input in module_inputs_torch_nn_TransformerEncoderLayer(
            None, device, dtype, requires_grad, training):
        # Construct a TransformerEncoderLayer object to pass to TransformerEncoder.
        l_args, l_kwargs = (layer_module_input.constructor_input.args,
                            layer_module_input.constructor_input.kwargs)
        encoder_layer = torch.nn.TransformerEncoderLayer(*l_args, **l_kwargs)
        num_layers = 2
        # Note: TransformerEncoderLayer takes a "src_mask" while
        # TransformerEncoder takes a "mask"; rename kwarg appropriately.
        forward_input = layer_module_input.forward_input
        if 'src_mask' in forward_input.kwargs:
            forward_input.kwargs['mask'] = forward_input.kwargs['src_mask']
            del forward_input.kwargs['src_mask']
        yield ModuleInput(
            constructor_input=FunctionInput(encoder_layer, num_layers),
            forward_input=forward_input,
            desc=layer_module_input.desc
        )

...

# ... in the module_db list ...
ModuleInfo(torch.nn.TransformerEncoder,
           train_and_eval_differ=True,
           module_inputs_func=module_inputs_torch_nn_TransformerEncoder,
           skips=(
               # No channels_last support for TransformerEncoderLayer currently.
               DecorateInfo(unittest.skip("Skipped!"), 'TestModule', 'test_memory_format'),
               # Doesn't support device / dtype kwargs directly because it is just a
               # container of TransformerEncoderLayers.
               DecorateInfo(unittest.expectedFailure, 'TestModule', 'test_factory_kwargs'),
               DecorateInfo(skipIfMps, 'TestModule', dtypes=[torch.float64]),)
           ),
```

File: test/functorch/test_aotdispatch.py [hunk 5]
```diff
@@ -2301,10 +2301,19 @@ def test_aot_autograd_symbolic_exhaustive(self, device, dtype, op):
...
+                                  # to a causal mask tensor, to see if Boolean is_causal should be set
+                                  # for TrnasformerEncoder layers, MHA and sdp custom kernels
+    torch.nn.Transformer,  # DataDependentOutputException: aten.equal compares a mask input
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply test/functorch/test_aotdispatch.py:2307]: Yes, it is - because we're now recognizing a causal mask and converting it to the is_causal boolean.  Since it depends on the mask input, it's input dependent evaluation.

File: torch/testing/_internal/common_modules.py (hunk 2 above)
Comment (mikekgfb) [reply torch/testing/_internal/common_modules.py:1404]: It calls TransformerEncoderLayer which differs according to pre-existing ModuleInfo entry, so I transitively bubbled up the flag.

File: torch/nn/modules/transformer.py [hunk 6]
```diff
@@ -287,13 +288,17 @@ def forward(
...
+                    torch.ones(sz, sz, device=mask.device) * float('-inf'), diagonal=1
+                ).to(mask.dtype)
+
+                if torch.equal(mask, causal_comparison):
+                    make_causal = True
 
-        if make_causal:
-            is_causal = True
-            mask = None
+        is_causal = make_causal
```
Comment (jbschlosser) [inline torch/nn/modules/transformer.py:301]: Are there tests for `is_causal`? I don't see any `ModuleInput` samples that set it.

Regardless, I don't think this is quite correct. If the caller explicitly sets `is_causal=True`, it will be overwritten here to be `False`.

File: torch/testing/_internal/common_modules.py (hunk 4 above)
Comment (mikekgfb) [reply torch/testing/_internal/common_modules.py:1424]: Thanks, this worked.

File: torch/nn/modules/transformer.py (hunk 6 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:301]: That can only happen if the attn_mask is not causal.  Arguably, that's the right outcome?

**Review by wushirong — APPROVED** [review]

File: torch/testing/_internal/common_modules.py (hunk 3 above)
Comment (mikekgfb) [reply torch/testing/_internal/common_modules.py]: Thanks!

File: test/test_transformers.py [hunk 7]
```diff
@@ -1330,6 +1330,34 @@ def test_flash_autocast_fp32_bfloat16(self):
                 _ = torch.nn.functional._scaled_dot_product_attention(
                     q, k, v, None, 0.0, False, False)
 
+    @parametrize("device", device_list)
+    @slowTest
+    def test_train_with_is_causal(self, device):
+        iters = 100
+        layer = nn.TransformerEncoderLayer(
+            d_model=2,
```
Comment (drisspg) [inline test/test_transformers.py:1338]: If we want to exercise the fused kernels I would change the d_model to be 16 hence 16/2 heads = 8.  Also, the failure initially occurred for fp16 right? Maybe we should do this test with that. 

Would this test fail prior to this PR? I am not 100% what its testing, besides more of an end-to-end exercise. Not sure what failures this shows though

File: torch/nn/functional.py [hunk 8]
```diff
@@ -4847,7 +4847,7 @@ def _in_projection(
      dropout_p (float): Dropout probability; if greater than 0.0, dropout is applied
      need_attn_weights (bool): If true, the second return value will contain the attention weights used;
          otherwise, the second return value is unspecified
-     is_causal (bool): If true, assumes causal attention masking; for this case, attn_mask should not be set.
+     is_causal (bool): If true, assumes causal attention masking and ignores attn_mask.
```
Comment (drisspg) [inline torch/nn/functional.py:4850]: Just a heads up I know there is a PR for xformers; https://github.com/facebookresearch/xformers/pull/587  that I think will allow for both casual plus additive attn bias. But I think for now this is the right way to go

File: torch/nn/modules/transformer.py [hunk 9]
```diff
@@ -286,14 +287,19 @@ def forward(
...
-            if is_causal:
-                raise RuntimeError("specify either mask or is_causal, but not both")
-
-        if make_causal:
-            is_causal = True
-            mask = None
+        make_causal = (is_causal is True)
+
+        if is_causal is None:
+            if mask is not None:
```
Comment (drisspg) [inline torch/nn/modules/transformer.py:293]: nit: if causal and mask is not None

**Review by drisspg — APPROVED** [review]
One small comment about the test but this looks good
